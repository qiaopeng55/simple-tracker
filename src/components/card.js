import React, { Component } from 'react';
import { DragSource, DropTarget } from 'react-dnd';
import { connect } from 'react-redux';
import { CARD } from '../actions/constant';
import { deleteCard, editingCard, dropCardOnCard } from '../actions/index';
import { StyledCard } from './styles';
import pencilEditButton from '../assets/pencil-edit-button.svg';
import deleteButton from '../assets/delete-button.svg';

/**
 * Implements the drag source contract.
 */
const cardDragSource = {
  beginDrag(props) {
    return {
      id: props.id,
      origColumnId: props.columnId,
    };
  },
};

const cardDropTarget = {
  drop(props, monitor) {
    const cardDragged = monitor.getItem();
    if (props.id !== cardDragged.id) {
      props.dropCardOnCard(cardDragged.id, props.id);
    }
  },
};

const collectDrag = (connectDnD, monitor) => ({
  connectDragSource: connectDnD.dragSource(),
  isDragging: monitor.isDragging(),
});

const collectDrop = (connectDnD, monitor) => ({
  connectDropTarget: connectDnD.dropTarget(),
  // You can ask the monitor about the current drag state:
  isOver: monitor.isOver(),
});

export class Card extends Component {
  render() {
    const {
      connectDragSource,
      connectDropTarget,
      id,
      text,
      columnId,
      isDragging,
      isOver,
      // eslint-disable-next-line no-shadow
      editingCard,
      // eslint-disable-next-line no-shadow
      deleteCard,
    } = this.props;
    return connectDropTarget(
      connectDragSource(
        <li>
          <StyledCard isDragging={isDragging} isOver={isOver}>
            <div className="task__title">{text}</div>
            <div
              className="action edit"
              onClick={() =>
                editingCard({
                  id,
                  text,
                  columnId,
                })
              }>
              <img src={pencilEditButton} alt="edit" />
            </div>
            <div className="action delete" onClick={() => deleteCard(id)}>
              <img src={deleteButton} alt="delete" />
            </div>
          </StyledCard>
        </li>
      )
    );
  }
}

const dragHighOrderCard = DragSource(CARD, cardDragSource, collectDrag)(Card);
const dragDropHighOrderCard = DropTarget(CARD, cardDropTarget, collectDrop)(
  dragHighOrderCard
);

export default connect(null, {
  dropCardOnCard,
  deleteCard,
  editingCard,
})(dragDropHighOrderCard);
