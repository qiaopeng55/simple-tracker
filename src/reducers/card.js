/**
 *  card reducer
 */

import uuidv1 from 'uuid/v1';
// Import Action Types
import {
  ADDING_CARD,
  UPDATE_CARD,
  SAVING_CARD,
  CLEAR_CARD,
  EDITING_CARD,
  EDITING_CARD_SAVE,
} from '../actions/constant';

function card(state = {}, action) {
  switch (action.type) {
    case ADDING_CARD:
      return {
        id: uuidv1(),
        columnId: action.data.columnId,
      };
    case UPDATE_CARD:
      return {
        ...action.data.card,
      };
    case SAVING_CARD:
    case CLEAR_CARD:
      return {};
    case EDITING_CARD:
      return {
        ...action.data.card,
      };
    case EDITING_CARD_SAVE:
      return {};
    default:
      return state;
  }
}

export default card;
