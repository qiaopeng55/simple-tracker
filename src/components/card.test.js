import React from 'react';
import { shallow } from 'enzyme';
import { Card } from './card';

describe('Card', () => {
  const mockDeleteCard = jest.fn();
  const mockEditingCard = jest.fn();

  const props = {
    id: 1,
    text: 'task 1',
    columnId: 1,
    deleteCard: mockDeleteCard,
    editingCard: mockEditingCard,
  };

  // Stub the React DnD connector functions with an identity function
  const identity = el => el;

  // Render with one set of props and test
  const card = shallow(
    <Card
      {...props}
      connectDragSource={identity}
      connectDropTarget={identity}
    />
  );

  it('renders properly', () => {
    expect(card).toMatchSnapshot();
  });

  it('renders task description properly', () => {
    expect(card.find('.task__title').text()).toEqual(props.text);
  });

  it('renders edit button', () => {
    expect(card.find('.action .edit').exists()).toEqual(true);
  });

  describe('when click edit button', () => {
    beforeEach(() => {
      card.find('.action .edit').simulate('click');
    });

    it('calls the editingCard callback with {id, text & columnId}', () => {
      const { id, text, columnId } = props;
      expect(mockEditingCard).toHaveBeenCalledWith({
        id,
        text,
        columnId,
      });
    });
  });

  it('renders delete button', () => {
    expect(card.find('.action .delete').exists()).toEqual(true);
  });

  describe('when click delete button', () => {
    beforeEach(() => {
      card.find('.action .delete').simulate('click');
    });

    it('calls the deleteCard callback with id', () => {
      const { id } = props;
      expect(mockDeleteCard).toHaveBeenCalledWith(id);
    });
  });
});
