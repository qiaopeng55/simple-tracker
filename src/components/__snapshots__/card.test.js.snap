// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders properly 1`] = `
<li>
  <styled.div>
    <div
      className="task__title"
    >
      task 1
    </div>
    <div
      className="action edit"
      onClick={[Function]}
    >
      <img
        alt="edit"
        src="pencil-edit-button.svg"
      />
    </div>
    <div
      className="action delete"
      onClick={[Function]}
    >
      <img
        alt="delete"
        src="delete-button.svg"
      />
    </div>
  </styled.div>
</li>
`;
